(DOCSP-27009) Add Realm React support to Open a Synced Realm (#2565)
mongodb/docs-realm: source/sdk/react-native/sync-data/configure-a-synced-realm.txt

sync in an App Services App, see :ref:`Atlas Device Sync Overview <sync>`.

Prerequisites
-------------

Before you open a realm with Flexible Sync in a React Native application:

#. :ref:`Configure Flexible Sync on the backend <enable-flexible-sync>`.
   You must configure Flexible Sync in the backend before you can use it with your
   client application.
#. :ref:`Initialize the App client <react-native-connect-to-mongodb-realm-backend-app>`.
#. :ref:`Authenticate a user <react-native-authenticate-users>` in your client project.

Open a Flexible Sync Realm
--------------------------

This section covers Flexible Sync realms. Flexible Sync is the preferred mode for
new apps that use :ref:`Atlas Device Sync <sync>`. 

For information about realms using the older Partition-Based Sync,
refer to :ref:`Partition-Based Sync <react-native-partition-realms>`.

By default, Realm syncs all data from the server before returning.
If you want to sync data in the background, read the :ref:`Open a Synced Realm
While Offline <react-native-open-synced-realm-offline>` section.

Nest <RealmProvider> within <AppProvider> and <UserProvider>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> To use a Flexible Sync realm and ``@realm/react``, you must nest ``<RealmProvider>``

Reviewer: ```suggestion
To use a Flexible Sync realm with ``@realm/react``, you must nest ``<RealmProvider>``
```

within ``<UserProvider>``. And ``<UserProvider>`` must be nested in ``<AppProvider>``. 
This ensures that ``<RealmProvider>`` has all of the context it needs, like a 
logged-in user.

> You can use ``<RealmProvider>`` by calling ``createRealmContext()`` with 
> your :js-sdk:`realm configuration <Realm.html#~Configuration>`, as shown in the 
> following example:

Reviewer: ```suggestion
Call ``createRealmContext()`` with 
your :js-sdk:`realm configuration <Realm.html#~Configuration>` to access ``RealmProvider`` in the return value.

The following example shows how to use these providers together in your app:
```

Reviewer: copy edit for clarity

   :language: javascript

For more information about configuring and using ``<AppProvider>``, check
out the :ref:`Connect To Atlas App Services page <react-native-app-client-configuration>`.

You can also find information about ``<UserProvider>`` on the
:ref:`Authenticate Users page <react-native-authenticate-users>`.

.. _react-native-flexible-sync-open-realm:

Configure a Flexible Sync Realm
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To configure a Flexible Sync realm, use ``@realm/react``'s ``createRealmContext()`` 
function and its returned ``<RealmProvider>``.

In a ``<RealmProvider>`` that's nested in a ``<UserProvider>``, add a ``sync`` 
property with a :js-sdk:`SyncConfiguration <Realm.App.Sync.html#~SyncConfiguration>` 
object that contains ``flexible: true``.

Note that ``<UserProvider>`` automatically passes an authenticated user
to ``<RealmProvider>``.

.. literalinclude:: /examples/generated/react-native/ts/configure-realm-sync.test.snippet.configure-realm-sync-full.tsx
   :language: javascript

.. important:: Flexible Sync Requires a Subscription

   You can't use a Flexible Sync realm until you add at least one subscription.

>    To learn how to add subscriptions, see: :ref:`<react-native-sync-add-subscription>`.

Reviewer: ```suggestion
   To learn how to add subscriptions, refer to :ref:`<react-native-sync-add-subscription>`.
```
